File: services/typescript/elysia/src/routes/handlers/put.ts
import type { Context } from "elysia";
import { type State, parseTask, parseId } from "app";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";

export function putHandler(state: State) {
  return ({ params, path, body, request, set }: Context) => {
    const logger = getLogger();

    const id = parseId(params.id);
    if (!id.success) {
      logger.warn({ method: request.method, path }, "Invalid path");
      return sendError(set, AppErrors.InvalidPath);
    }

    const task = parseTask(body);
    if (!task.success) {
      logger.warn({ method: request.method, path }, "Invalid body JSON");
      return sendError(set, AppErrors.InvalidJsonBody);
    }

    const previous_task = state.tasks.get(id.data);

    if (previous_task === undefined) {
      logger.warn({ id: id.data, method: "PUT", path }, "Task not found");
      return sendError(set, AppErrors.TaskNotFound);
    }

    state.tasks.set(id.data, task.data);

    logger.info(
      {
        id: id.data,
        from_secret: previous_task.secret.length,
        to_secret: task.data.secret.length,
        from_operation: previous_task.operation.toString().toLowerCase(),
        to_operation: task.data.operation.toString().toLowerCase(),
        method: "PUT"
      },
      "Overwrote task"
    );

    set.status = 200;
    return task.data;
  };
}

File: services/python/django/src/routes/handlers/put.py
import structlog
import json
from uuid import UUID
from pydantic import ValidationError
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.apps import apps
from typing import cast

from app.params import parse_id, parse_task
from app.task import Task
from routes.errors import send_error, AppErrors
from routes.apps import RoutesConfig


async def put_handler(request: HttpRequest, id: str) -> HttpResponse:
    logger = structlog.get_logger()

    try:
        task_id: UUID = parse_id(id)
    except ValidationError:
        logger.warn("Invalid path", method=request.method, path=request.path)
        return send_error(AppErrors.InvalidPath)

    try:
        body = json.loads(request.body)
    except (json.JSONDecodeError, UnicodeDecodeError):
        body = None

    if body is None:
        logger.warn("Invalid body JSON", method=request.method, path=request.path)
        return send_error(AppErrors.InvalidJsonBody)

    try:
        new_task: Task = parse_task(body)
    except ValidationError:
        logger.warn("Invalid body JSON", method=request.method, path=request.path)
        return send_error(AppErrors.InvalidJsonBody)

    config = cast(RoutesConfig, apps.get_app_config("routes"))

    async with config.tasks_lock:
        task: Task | None = config.state.tasks.get(task_id)

        if task is None:
            logger.warn(
                "Task not found",
                id=str(task_id),
                method=request.method,
                path=request.path,
            )
            return send_error(AppErrors.TaskNotFound)

        config.state.tasks[task_id] = new_task

    logger.info(
        "Overwrote task",
        id=str(task_id),
        from_secret=len(task.secret),
        to_secret=len(new_task.secret),
        from_operation=task.operation.lower(),
        to_operation=new_task.operation.lower(),
        method=request.method,
    )

    return JsonResponse(new_task.model_dump(), status=200)
